File: bot.py
import discord
from discord.ext import commands
from discord import app_commands, ui
import os
from dotenv import load_dotenv
import asyncio

# 環境変数の読み込み
load_dotenv()
TOKEN = os.getenv('DISCORD_TOKEN')
# 募集内容を表示するチャンネルID (未設定の場合は環境変数から読み込むか、コード内で指定)
TARGET_CHANNEL_ID = os.getenv('TARGET_CHANNEL_ID')

if not TOKEN:
    print("エラー: .envファイルにDISCORD_TOKENを設定してください。")
    exit()

# インテントの設定
intents = discord.Intents.default()
intents.message_content = True
intents.guilds = True
intents.members = True

# ボットのセットアップ
class FF14RecruitBot(commands.Bot):
    def __init__(self):
        super().__init__(command_prefix='!', intents=intents)

    async def setup_hook(self):
        # 永続的なViewを登録（再起動後もボタンが動くようにするため）
        self.add_view(RecruitTriggerView())
        # 既存の募集用Viewも本来はDBから復元してadd_viewすべきですが、
        # 簡易実装のため、起動後に作成されたものは動きますが、再起動前のボタンは無効になる場合があります。
        # 本格運用ではDBが必要です。
        await self.tree.sync()

bot = FF14RecruitBot()

# ---------------------------------------------------------
# データ管理 (簡易的なメモリ内保存)
# 本格運用にはSQLiteなどのデータベース推奨
# ---------------------------------------------------------
recruitments = {}

class RecruitmentData:
    def __init__(self, host_id, title, date, time, max_roles, vc_id=None):
        self.host_id = host_id
        self.title = title
        self.date = date
        self.time = time
        self.max_roles = max_roles  # {'Tank': 2, 'Healer': 2, 'DPS': 4}
        self.current_roles = {'Tank': [], 'Healer': [], 'DPS': []} # IDのリスト
        self.vc_id = vc_id
        self.is_active = True

    def add_participant(self, user_id, role):
        # 既に参加しているかチェックして削除（ロール変更対応）
        self.remove_participant(user_id)
        
        if len(self.current_roles[role]) < self.max_roles[role]:
            self.current_roles[role].append(user_id)
            return True
        return False

    def remove_participant(self, user_id):
        for role in self.current_roles:
            if user_id in self.current_roles[role]:
                self.current_roles[role].remove(user_id)
                return True
        return False

    def get_status_str(self, role):
        return f"{len(self.current_roles[role])}/{self.max_roles[role]}"

    def is_full(self, role):
        return len(self.current_roles[role]) >= self.max_roles[role]

# ---------------------------------------------------------
# UIコンポーネント: 募集操作パネル (参加・キャンセル・終了)
# ---------------------------------------------------------
class RecruitmentPanel(ui.View):
    def __init__(self, message_id):
        super().__init__(timeout=None)
        self.message_id = message_id
        self.update_buttons()

    def get_data(self):
        return recruitments.get(self.message_id)

    def update_buttons(self):
        data = self.get_data()
        if not data:
            self.clear_items()
            return

        # ボタンの定義を動的に更新
        # Tank
        tank_btn = [x for x in self.children if isinstance(x, ui.Button) and x.custom_id == f"role_tank_{self.message_id}"]
        if not tank_btn:
            self.add_item(RoleButton(role="Tank", message_id=self.message_id))
        
        # Healer
        healer_btn = [x for x in self.children if isinstance(x, ui.Button) and x.custom_id == f"role_healer_{self.message_id}"]
        if not healer_btn:
            self.add_item(RoleButton(role="Healer", message_id=self.message_id))

        # DPS
        dps_btn = [x for x in self.children if isinstance(x, ui.Button) and x.custom_id == f"role_dps_{self.message_id}"]
        if not dps_btn:
            self.add_item(RoleButton(role="DPS", message_id=self.message_id))

        # Cancel & Close (これらは固定で追加済みだが、ラベル更新が必要な場合はここで処理)

class RoleButton(ui.Button):
    def __init__(self, role, message_id):
        self.role = role
        self.message_id = message_id
        super().__init__(
            style=discord.ButtonStyle.primary if role == "Tank" else discord.ButtonStyle.success if role == "Healer" else discord.ButtonStyle.danger,
            label=f"{role} ?",
            custom_id=f"role_{role.lower()}_{message_id}",
            row=0
        )
        # 初期化時にラベルを設定
        self.refresh_label()

    def refresh_label(self):
        data = recruitments.get(self.message_id)
        if data:
            count = len(data.current_roles[self.role])
            max_c = data.max_roles[self.role]
            self.label = f"{self.role} {count}/{max_c}"
            self.disabled = count >= max_c

    async def callback(self, interaction: discord.Interaction):
        data = recruitments.get(self.message_id)
        if not data:
            await interaction.response.send_message("この募集は既に終了しているか、データが見つかりません。", ephemeral=True)
            return

        # 参加処理
        success = data.add_participant(interaction.user.id, self.role)
        
        if success:
            await update_recruitment_message(interaction, self.message_id)
            await interaction.followup.send(f"{self.role}枠に参加しました！", ephemeral=True)
        else:
            await interaction.response.send_message("その枠は満員です。", ephemeral=True)

class LeaveButton(ui.Button):
    def __init__(self, message_id):
        super().__init__(style=discord.ButtonStyle.secondary, label="参加取消", custom_id=f"leave_{message_id}", row=1)
        self.message_id = message_id

    async def callback(self, interaction: discord.Interaction):
        data = recruitments.get(self.message_id)
        if not data:
            await interaction.response.send_message("データが見つかりません。", ephemeral=True)
            return

        if data.remove_participant(interaction.user.id):
            await update_recruitment_message(interaction, self.message_id)
            await interaction.followup.send("参加を取り消しました。", ephemeral=True)
        else:
            await interaction.response.send_message("参加していません。", ephemeral=True)

class CloseButton(ui.Button):
    def __init__(self, message_id):
        super().__init__(style=discord.ButtonStyle.danger, label="募集終了", custom_id=f"close_{message_id}", row=1)
        self.message_id = message_id

    async def callback(self, interaction: discord.Interaction):
        data = recruitments.get(self.message_id)
        if not data:
            await interaction.response.send_message("データが見つかりません。", ephemeral=True)
            return

        if interaction.user.id != data.host_id:
            await interaction.response.send_message("募集主のみが終了できます。", ephemeral=True)
            return

        # VC削除
        if data.vc_id:
            vc_channel = interaction.guild.get_channel(data.vc_id)
            if vc_channel:
                try:
                    await vc_channel.delete()
                except:
                    pass
        
        # メッセージ削除
        try:
            await interaction.message.delete()
        except:
            pass
        
        # データ削除
        del recruitments[self.message_id]
        await interaction.response.send_message("募集を終了し、削除しました。", ephemeral=True)

# ---------------------------------------------------------
# メッセージ更新関数
# ---------------------------------------------------------
async def update_recruitment_message(interaction: discord.Interaction, message_id):
    # ボタン操作への応答としてメッセージを更新するため、deferが必要な場合がある
    if not interaction.response.is_done():
        await interaction.response.defer()

    data = recruitments.get(message_id)
    if not data:
        return

    # Embedの作り直し
    embed = discord.Embed(title=f"募集: {data.title}", color=discord.Color.blue())
    embed.add_field(name="開催日時", value=f"{data.date} {data.time}", inline=False)
    embed.add_field(name="募集主", value=f"<@{data.host_id}>", inline=False)
    
    # 参加者リストの生成
    for role in ["Tank", "Healer", "DPS"]:
        members = data.current_roles[role]
        member_str = "\n".join([f"<@{uid}>" for uid in members]) if members else "なし"
        embed.add_field(name=f"{role} ({len(members)}/{data.max_roles[role]})", value=member_str, inline=True)

    if data.vc_id:
        embed.add_field(name="VC", value=f"<#{data.vc_id}>", inline=False)

    # Viewの再構築（ボタンのラベル更新のため）
    new_view = ui.View(timeout=None)
    
    # 各ロールボタン
    for role in ["Tank", "Healer", "DPS"]:
        btn = RoleButton(role, message_id)
        btn.refresh_label()
        new_view.add_item(btn)
    
    new_view.add_item(LeaveButton(message_id))
    new_view.add_item(CloseButton(message_id))

    try:
        msg = await interaction.channel.fetch_message(message_id)
        await msg.edit(embed=embed, view=new_view)
    except discord.NotFound:
        pass

# ---------------------------------------------------------
# UIコンポーネント: 入力フォーム (Modal)
# ---------------------------------------------------------
class RecruitModal(ui.Modal, title='募集内容の設定'):
    def __init__(self, use_vc=False):
        super().__init__()
        self.use_vc = use_vc
        # TextInputの定義をここで行う
        self.title_input = ui.TextInput(label='タイトル', placeholder='エデン零式 1層練習', max_length=50)
        self.datetime_input = ui.TextInput(label='開催日時 (例: 1201 21:00)', placeholder='20231201 21:00', min_length=5, max_length=20)
        self.tank_input = ui.TextInput(label='Tank募集人数', placeholder='2', min_length=1, max_length=2, default='2')
        self.healer_input = ui.TextInput(label='Healer募集人数', placeholder='2', min_length=1, max_length=2, default='2')
        self.dps_input = ui.TextInput(label='DPS募集人数', placeholder='4', min_length=1, max_length=2, default='4')

        self.add_item(self.title_input)
        self.add_item(self.datetime_input)
        self.add_item(self.tank_input)
        self.add_item(self.healer_input)
        self.add_item(self.dps_input)

    async def on_submit(self, interaction: discord.Interaction):
        # 入力値の取得
        title = self.title_input.value
        datetime_val = self.datetime_input.value
        
        # 日時を分割（簡易的）
        parts = datetime_val.split()
        if len(parts) >= 2:
            date_val = parts[0]
            time_val = parts[1]
        else:
            date_val = datetime_val
            time_val = ""

        # ロール人数のパース
        try:
            t = int(self.tank_input.value)
            h = int(self.healer_input.value)
            d = int(self.dps_input.value)
            max_roles = {'Tank': t, 'Healer': h, 'DPS': d}
        except ValueError:
            await interaction.response.send_message("人数の形式が正しくありません。半角数字で入力してください。", ephemeral=True)
            return

        # ターゲットチャンネルの取得
        target_channel_id = TARGET_CHANNEL_ID
        if not target_channel_id:
            target_channel = interaction.channel
        else:
            target_channel = interaction.guild.get_channel(int(target_channel_id))
            if not target_channel:
                await interaction.response.send_message("募集チャンネルが見つかりません。", ephemeral=True)
                return

        # VC作成
        vc_id = None
        if self.use_vc:
            guild = interaction.guild
            vc_name = f"🔑_{title}_VC" 
            try:
                vc = await guild.create_voice_channel(name=vc_name)
                vc_id = vc.id
            except Exception as e:
                await interaction.response.send_message(f"VC作成に失敗しました: {e}", ephemeral=True)
                return

        # 仮メッセージ送信（ID確保のため）
        embed = discord.Embed(title="募集中...", description="準備中")
        msg = await target_channel.send(embed=embed)

        # データ保存
        data = RecruitmentData(
            host_id=interaction.user.id,
            title=title,
            date=date_val,
            time=time_val,
            max_roles=max_roles,
            vc_id=vc_id
        )
        recruitments[msg.id] = data

        # Viewを作成してメッセージ更新
        view = ui.View(timeout=None)
        
        for role in ["Tank", "Healer", "DPS"]:
            btn = RoleButton(role, msg.id)
            btn.refresh_label()
            view.add_item(btn)
        
        view.add_item(LeaveButton(msg.id))
        view.add_item(CloseButton(msg.id))

        # 初回のEmbed更新
        await update_recruitment_message(interaction, msg.id)
        
        # Modalへの応答
        if not interaction.response.is_done():
            await interaction.response.send_message(f"募集を作成しました！ -> {msg.jump_url}", ephemeral=True)
        else:
            await interaction.followup.send(f"募集を作成しました！ -> {msg.jump_url}", ephemeral=True)

# ---------------------------------------------------------
# UIコンポーネント: 募集作成トリガー (常設ボタン)
# ---------------------------------------------------------
class RecruitTriggerView(ui.View):
    def __init__(self):
        super().__init__(timeout=None)

    @ui.button(label="募集を作成", style=discord.ButtonStyle.primary, custom_id="trigger_create_recruit_no_vc")
    async def create_recruit_no_vc(self, interaction: discord.Interaction, button: ui.Button):
        await interaction.response.send_modal(RecruitModal(use_vc=False))

    @ui.button(label="募集を作成 (+VC)", style=discord.ButtonStyle.secondary, custom_id="trigger_create_recruit_with_vc")
    async def create_recruit_with_vc(self, interaction: discord.Interaction, button: ui.Button):
        await interaction.response.send_modal(RecruitModal(use_vc=True))

# ---------------------------------------------------------
# コマンド
# ---------------------------------------------------------
@bot.event
async def on_ready():
    print(f'Logged in as {bot.user} (ID: {bot.user.id})')
    print('------')

@bot.command()
async def setup(ctx):
    """
    募集作成ボタンを設置するコマンド
    """
    embed = discord.Embed(
        title="パーティー募集",
        description="下のボタンを押して募集を開始してください。",
        color=discord.Color.gold()
    )
    await ctx.send(embed=embed, view=RecruitTriggerView())

bot.run(TOKEN)

File: src/bot.ts
import {
  createBot,
  Intents,
  startBot,
  CreateMessage,
  Embed,
  MessageComponents,
  InteractionResponseTypes,
  ApplicationCommandTypes,
  InteractionTypes,
  ButtonStyles,
  MessageComponentTypes,
  ChannelTypes,
} from "https://deno.land/x/discordeno@18.0.1/mod.ts";
import { load } from "https://deno.land/std@0.208.0/dotenv/mod.ts";

// 環境変数の読み込み
await load({ export: true });

const TOKEN = Deno.env.get("DISCORD_TOKEN");
const TARGET_CHANNEL_ID = Deno.env.get("TARGET_CHANNEL_ID");

// 定義されていないEnumを手動定義
const TextInputStyles = {
  Short: 1,
  Paragraph: 2,
} as const;

if (!TOKEN) {
  console.error("エラー: DISCORD_TOKENが設定されていません。");
  // Deno Deployの環境変数が設定されているか確認してください
}

// ---------------------------------------------------------
// データ定義
// ---------------------------------------------------------

interface RecruitmentData {
  hostId: bigint;
  title: string;
  date: string;
  time: string;
  maxRoles: {
    Tank: number;
    Healer: number;
    DPS: number;
  };
  currentRoles: {
    Tank: bigint[];
    Healer: bigint[];
    DPS: bigint[];
  };
  vcId: bigint | null;
}

// 簡易的なメモリ内保存
const recruitments = new Map<string, RecruitmentData>();

// ---------------------------------------------------------
// ユーティリティ関数
// ---------------------------------------------------------

// 募集パネル(Embed + Buttons)の更新
async function updateRecruitmentMessage(bot: any, channelId: bigint, messageId: bigint) {
  const data = recruitments.get(messageId.toString());
  if (!data) return;

  // Embedの作成
  const embed: Embed = {
    title: `募集: ${data.title}`,
    color: 0x0099ff,
    fields: [
      { name: "開催日時", value: `${data.date} ${data.time}`, inline: false },
      { name: "募集主", value: `<@${data.hostId}>`, inline: false },
    ],
  };

  // 参加者リスト
  const roles: ("Tank" | "Healer" | "DPS")[] = ["Tank", "Healer", "DPS"];
  for (const role of roles) {
    const members = data.currentRoles[role];
    const memberStr = members.length > 0
      ? members.map((uid) => `<@${uid}>`).join("\n")
      : "なし";
    embed.fields!.push({
      name: `${role} (${members.length}/${data.maxRoles[role]})`,
      value: memberStr,
      inline: true,
    });
  }

  if (data.vcId) {
    embed.fields!.push({ name: "VC", value: `<#${data.vcId}>`, inline: false });
  }

  // ボタンの作成
  const components: MessageComponents = [];
  
  // Row 1: Role Buttons
  const row1 = {
    type: MessageComponentTypes.ActionRow,
    components: [] as any[],
  };

  for (const role of roles) {
    const count = data.currentRoles[role].length;
    const max = data.maxRoles[role];
    const isFull = count >= max;

    let style = ButtonStyles.Primary;
    if (role === "Healer") style = ButtonStyles.Success;
    if (role === "DPS") style = ButtonStyles.Danger;

    row1.components.push({
      type: MessageComponentTypes.Button,
      customId: `role_${role.toLowerCase()}_${messageId}`,
      label: `${role} ${count}/${max}`,
      style: style,
      disabled: isFull,
    });
  }
  components.push(row1);

  // Row 2: Control Buttons
  const row2 = {
    type: MessageComponentTypes.ActionRow,
    components: [
      {
        type: MessageComponentTypes.Button,
        customId: `leave_${messageId}`,
        label: "参加取消",
        style: ButtonStyles.Secondary,
      },
      {
        type: MessageComponentTypes.Button,
        customId: `close_${messageId}`,
        label: "募集終了",
        style: ButtonStyles.Danger,
      },
    ],
  };
  components.push(row2);

  try {
    await bot.helpers.editMessage(channelId, messageId, {
      embeds: [embed],
      components: components,
    });
  } catch (error) {
    console.error("メッセージ更新エラー:", error);
  }
}

// ---------------------------------------------------------
// Bot作成
// ---------------------------------------------------------

const bot = createBot({
  token: TOKEN || "", // TOKENがない場合は空文字で初期化し、startBotでエラーになるようにする
  intents: Intents.Guilds | Intents.GuildMessages | Intents.MessageContent | Intents.GuildMembers | Intents.GuildVoiceStates,
  events: {
    ready: (_bot, payload) => {
      console.log(`${payload.user.username} is ready!`);
    },
    // !setup コマンド
    messageCreate: async (bot, message) => {
      if (message.isBot) return;

      if (message.content === "!setup") {
        const embed: Embed = {
          title: "パーティー募集",
          description: "下のボタンを押して募集を開始してください。",
          color: 0xFFD700,
        };

        const components: MessageComponents = [{
          type: MessageComponentTypes.ActionRow,
          components: [
            {
              type: MessageComponentTypes.Button,
              customId: "trigger_create_recruit_no_vc",
              label: "募集を作成",
              style: ButtonStyles.Primary,
            },
            {
              type: MessageComponentTypes.Button,
              customId: "trigger_create_recruit_with_vc",
              label: "募集を作成 (+VC)",
              style: ButtonStyles.Secondary,
            },
          ],
        }];

        await bot.helpers.sendMessage(message.channelId, {
          embeds: [embed],
          components: components,
        });
      }
    },
    // インタラクション処理
    interactionCreate: async (bot, interaction) => {
        try {
            // 1. 募集作成ボタン -> モーダル表示
            if (interaction.type === InteractionTypes.MessageComponent && interaction.data?.componentType === MessageComponentTypes.Button) {
                if (interaction.data.customId === 'trigger_create_recruit_no_vc' || interaction.data.customId === 'trigger_create_recruit_with_vc') {
                    const useVc = interaction.data.customId === 'trigger_create_recruit_with_vc';
                    
                    await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                        type: InteractionResponseTypes.Modal,
                        data: {
                            customId: `modal_recruit_create_${useVc ? 'vc' : 'novc'}`,
                            title: "募集内容の設定",
                            components: [
                                {
                                    type: MessageComponentTypes.ActionRow,
                                    components: [{
                                        type: MessageComponentTypes.InputText,
                                        customId: "title_input",
                                        style: TextInputStyles.Short,
                                        label: "タイトル",
                                        placeholder: "エデン零式 1層練習",
                                        maxLength: 50
                                    }]
                                },
                                {
                                    type: MessageComponentTypes.ActionRow,
                                    components: [{
                                        type: MessageComponentTypes.InputText,
                                        customId: "datetime_input",
                                        style: TextInputStyles.Short,
                                        label: "開催日時 (例: 1201 21:00)",
                                        placeholder: "20231201 21:00",
                                        minLength: 5,
                                        maxLength: 20
                                    }]
                                },
                                {
                                    type: MessageComponentTypes.ActionRow,
                                    components: [{
                                        type: MessageComponentTypes.InputText,
                                        customId: "tank_input",
                                        style: TextInputStyles.Short,
                                        label: "Tank募集人数",
                                        placeholder: "2",
                                        value: "2",
                                        minLength: 1,
                                        maxLength: 2
                                    }]
                                },
                                {
                                    type: MessageComponentTypes.ActionRow,
                                    components: [{
                                        type: MessageComponentTypes.InputText,
                                        customId: "healer_input",
                                        style: TextInputStyles.Short,
                                        label: "Healer募集人数",
                                        placeholder: "2",
                                        value: "2",
                                        minLength: 1,
                                        maxLength: 2
                                    }]
                                },
                                {
                                    type: MessageComponentTypes.ActionRow,
                                    components: [{
                                        type: MessageComponentTypes.InputText,
                                        customId: "dps_input",
                                        style: TextInputStyles.Short,
                                        label: "DPS募集人数",
                                        placeholder: "4",
                                        value: "4",
                                        minLength: 1,
                                        maxLength: 2
                                    }]
                                }
                            ]
                        }
                    });
                    return;
                }
            }

            // 2. モーダル提出
            if (interaction.type === InteractionTypes.ModalSubmit) {
                if (interaction.data?.customId?.startsWith('modal_recruit_create_')) {
                    const useVc = interaction.data.customId.endsWith('_vc');
                    
                    // Discordenoではcomponentsの構造が少し異なるため、findで取得
                    const getVal = (id: string) => {
                        // モーダルのコンポーネントはActionRowの中に入っている
                        for (const row of interaction.data?.components || []) {
                            const comp = row.components?.find(c => c.customId === id);
                            if (comp) return comp.value || "";
                        }
                        return "";
                    };

                    const title = getVal('title_input');
                    const datetimeVal = getVal('datetime_input');
                    const tStr = getVal('tank_input');
                    const hStr = getVal('healer_input');
                    const dStr = getVal('dps_input');

                    const t = parseInt(tStr);
                    const h = parseInt(hStr);
                    const d = parseInt(dStr);

                    if (isNaN(t) || isNaN(h) || isNaN(d)) {
                        await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                            type: InteractionResponseTypes.ChannelMessageWithSource,
                            data: { content: '人数は半角数字で入力してください。', flags: 64 } // ephemeral
                        });
                        return;
                    }

                    const parts = datetimeVal.split(/\s+/);
                    const dateVal = parts[0];
                    const timeVal = parts.length > 1 ? parts[1] : '';

                    // ターゲットチャンネル
                    let targetChannelId = interaction.channelId!;
                    if (TARGET_CHANNEL_ID) {
                        targetChannelId = BigInt(TARGET_CHANNEL_ID);
                    }

                    // VC作成
                    let vcId: bigint | null = null;
                    if (useVc && interaction.guildId) {
                        const vcName = `🔑_${title}_VC`;
                        try {
                            const vc = await bot.helpers.createChannel(interaction.guildId, {
                                name: vcName,
                                type: ChannelTypes.GuildVoice,
                            });
                            vcId = vc.id;
                        } catch (e) {
                             await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                                type: InteractionResponseTypes.ChannelMessageWithSource,
                                data: { content: `VC作成に失敗しました: ${e}`, flags: 64 }
                            });
                            return;
                        }
                    }

                    // 仮メッセージ送信 (InteractionResponseではなく通常のメッセージとして送信してIDを取得する)
                    // まずはInteractionへの応答を返す（読み込み中...などを消すため）
                    await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                        type: InteractionResponseTypes.DeferredChannelMessageWithSource,
                        data: { flags: 64 }
                    });

                    const embed: Embed = { title: "募集中...", description: "準備中" };
                    const msg = await bot.helpers.sendMessage(targetChannelId, { embeds: [embed] });

                    // データ保存
                    const data: RecruitmentData = {
                        hostId: interaction.user.id,
                        title: title,
                        date: dateVal,
                        time: timeVal,
                        maxRoles: { Tank: t, Healer: h, DPS: d },
                        currentRoles: { Tank: [], Healer: [], DPS: [] },
                        vcId: vcId
                    };
                    recruitments.set(msg.id.toString(), data);

                    // メッセージ更新
                    await updateRecruitmentMessage(bot, targetChannelId, msg.id);

                    // Deferred応答への追記
                    await bot.helpers.editOriginalInteractionResponse(interaction.token, {
                        content: `募集を作成しました！ -> https://discord.com/channels/${interaction.guildId}/${targetChannelId}/${msg.id}`
                    });
                    return;
                }
            }

            // 3. ボタン操作
            if (interaction.type === InteractionTypes.MessageComponent && interaction.data?.componentType === MessageComponentTypes.Button) {
                const parts = interaction.data.customId!.split('_');
                if (parts.length < 2) return;

                const action = parts[0];
                const messageId = parts[parts.length - 1];

                const data = recruitments.get(messageId);
                if (!data) {
                    if (['close', 'leave', 'role'].includes(action)) {
                        await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                            type: InteractionResponseTypes.ChannelMessageWithSource,
                            data: { content: 'この募集データは見つかりません（再起動された可能性があります）。', flags: 64 }
                        });
                    }
                    return;
                }

                if (action === 'role') {
                    const roleKey = parts[1];
                    const roleMap: {[key: string]: 'Tank' | 'Healer' | 'DPS'} = {
                        'tank': 'Tank', 'healer': 'Healer', 'dps': 'DPS'
                    };
                    const role = roleMap[roleKey];
                    if (!role) return;

                    // ロール参加処理
                    let removed = false;
                    ['Tank', 'Healer', 'DPS'].forEach((r) => {
                        const rKey = r as 'Tank' | 'Healer' | 'DPS';
                        if (data.currentRoles[rKey].includes(interaction.user.id)) {
                            data.currentRoles[rKey] = data.currentRoles[rKey].filter(uid => uid !== interaction.user.id);
                            removed = true;
                        }
                    });

                    if (data.currentRoles[role].length >= data.maxRoles[role]) {
                        if (removed) {
                             // 巻き戻し処理が必要だが省略
                        }
                        await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                            type: InteractionResponseTypes.ChannelMessageWithSource,
                            data: { content: 'その枠は満員です。', flags: 64 }
                        });
                        return;
                    }

                    data.currentRoles[role].push(interaction.user.id);
                    
                    // 更新処理
                    await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                        type: InteractionResponseTypes.DeferredUpdateMessage,
                    });
                    await updateRecruitmentMessage(bot, interaction.channelId!, BigInt(messageId));
                    
                    // 完了通知はephemeralメッセージで送るか、更新だけで済ますか。今回は更新だけ。
                    // await bot.helpers.sendFollowupMessage(interaction.token, { content: `${role}枠に参加しました！`, flags: 64 });

                } else if (action === 'leave') {
                    let removed = false;
                    ['Tank', 'Healer', 'DPS'].forEach((r) => {
                        const rKey = r as 'Tank' | 'Healer' | 'DPS';
                        if (data.currentRoles[rKey].includes(interaction.user.id)) {
                            data.currentRoles[rKey] = data.currentRoles[rKey].filter(uid => uid !== interaction.user.id);
                            removed = true;
                        }
                    });

                    if (removed) {
                        await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                            type: InteractionResponseTypes.DeferredUpdateMessage,
                        });
                        await updateRecruitmentMessage(bot, interaction.channelId!, BigInt(messageId));
                    } else {
                        await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                            type: InteractionResponseTypes.ChannelMessageWithSource,
                            data: { content: '参加していません。', flags: 64 }
                        });
                    }

                } else if (action === 'close') {
                    if (interaction.user.id !== data.hostId) {
                         await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                            type: InteractionResponseTypes.ChannelMessageWithSource,
                            data: { content: '募集主のみが終了できます。', flags: 64 }
                        });
                        return;
                    }

                    if (data.vcId && interaction.guildId) {
                        try {
                            await bot.helpers.deleteChannel(data.vcId);
                        } catch (e) {
                            console.error("VC削除エラー", e);
                        }
                    }

                    try {
                        await bot.helpers.deleteMessage(interaction.channelId!, BigInt(messageId));
                    } catch (e) {
                        console.error("メッセージ削除エラー", e);
                    }

                    recruitments.delete(messageId);
                    await bot.helpers.sendInteractionResponse(interaction.id, interaction.token, {
                        type: InteractionResponseTypes.ChannelMessageWithSource,
                        data: { content: '募集を終了し、削除しました。', flags: 64 }
                    });
                }
            }

        } catch (err) {
            console.error("Interaction Error:", err);
        }
    }
  },
});

Deno.cron("Continuous Request", "*/2 * * * *", () => {
    console.log("running...");
});

if (TOKEN) {
    await startBot(bot);
} else {
    console.error("Bot token not found. Please set DISCORD_TOKEN in your environment variables.");
}
